libsolidity/ast/ASTUtils.h

```diff
@@ -48,4 +51,9 @@ Type const* type(Expression const& _expression);
 /// (this can happen for variables with non-explicit types before their types are resolved)
 Type const* type(VariableDeclaration const& _variable);
 
+/// @returns the value of the base slot of the contract's inheritance hierarchy storage if it is explicitly specified.
+/// In any other case, including invalid ones, returns the default value of zero.
```

Comment:
```suggestion
/// @returns The base slot of the inheritance hierarchy rooted at the specified contract.
/// The value comes from the inheritance specifier of the contract and defaults to zero.
/// The value is zero also when the contract is an interface or a library (and cannot have storage).
```

BTW, I saw that you initially had an assert against interfaces and libraries and I thought that was a good idea. Better than returning a zero and pretending they have storage, which may easily lead to mistakes. Some tests were failing, but you probably just need an `if` in `StorageOffsets` to handle that.

Reply:
Yeah, I removed the asserts to get tests green and was planning to fix it later. Should we include abstract contracts in the restrictios too?

Reply:
Abstract contracts definitely should be allowed. Those can have storage variables and they are reported to start at 0 if you request `--storage-layout`.

The reason to exclude libraries and interfaces would be because for these even asking about the layout makes no sense. But looking at `ContractType::stateVariables()` now, I see that we actually do ask that question and `--storage-layout` even reports an empty layout for them (rather than giving no output). If that's the case I'd remove these assertions after all. My assumption was that for libraries and interfaces you wouldn't even run `computeOffsets()` at all and wouldn't need to use any value for the base slot in the first place. But you still do so it's better for `layoutBaseForInheritanceHierarchy()` to return it rather than hard-code it inside `stateVariables()`.

Reply:
Please update the docstring.